### PR TITLE
(mostly) copy from huntsman v2 analog for huntsman mini analog

### PR DESCRIPTION
**Pull request by meeuw**
I hesitated for too long and just bought the Razer Huntsman Mini :tada: (I like it so far!) https://github.com/openrazer/openrazer/issues/1579#issuecomment-1176611245

I'm guessing it's similar to the Huntsman V2 Analog and (I think) I successfully added the same snippets as for the Huntsman V2 Analog for the Huntsman Mini.

I compiled the kernel module, inserted it into the kernel:

```
input: Razer Razer Huntsman Mini Analog as /devices/pci0000:00/0000:00:14.0/usb3/3-6/3-6.1/3-6.1.4/3-6.1.4:1.0/0003:1532:0282.0006/input/input31
hid-generic 0003:1532:0282.0006: input,hidraw1: USB HID v1.11 Keyboard [Razer Razer Huntsman Mini Analog] on usb-0000:00:14.0-6.1.4/input0
input: Razer Razer Huntsman Mini Analog Keyboard as /devices/pci0000:00/0000:00:14.0/usb3/3-6/3-6.1/3-6.1.4/3-6.1.4:1.1/0003:1532:0282.0007/input/input32
input: Razer Razer Huntsman Mini Analog as /devices/pci0000:00/0000:00:14.0/usb3/3-6/3-6.1/3-6.1.4/3-6.1.4:1.1/0003:1532:0282.0007/input/input33
hid-generic 0003:1532:0282.0007: input,hidraw3: USB HID v1.11 Keyboard [Razer Razer Huntsman Mini Analog] on usb-0000:00:14.0-6.1.4/input1
input: Razer Razer Huntsman Mini Analog as /devices/pci0000:00/0000:00:14.0/usb3/3-6/3-6.1/3-6.1.4/3-6.1.4:1.2/0003:1532:0282.0008/input/input34
hid-generic 0003:1532:0282.0008: input,hidraw4: USB HID v1.11 Mouse [Razer Razer Huntsman Mini Analog] on usb-0000:00:14.0-6.1.4/input2
hid-generic 0003:1532:0282.0009: No inputs registered, leaving
hid-generic 0003:1532:0282.0009: hidraw5: USB HID v1.11 Device [Razer Razer Huntsman Mini Analog] on usb-0000:00:14.0-6.1.4/input3
input: Razer Razer Huntsman Mini Analog Consumer Control as /devices/pci0000:00/0000:00:14.0/usb3/3-6/3-6.1/3-6.1.4/3-6.1.4:1.4/0003:1532:0282.000A/input/input35
hid-generic 0003:1532:0282.000A: input,hiddev97,hidraw6: USB HID v1.11 Device [Razer Razer Huntsman Mini Analog] on usb-0000:00:14.0-6.1.4/input4
input: Razer Razer Huntsman Mini Analog as /devices/pci0000:00/0000:00:14.0/usb3/3-6/3-6.1/3-6.1.4/3-6.1.4:1.0/0003:1532:0282.0006/input/input36
razerkbd 0003:1532:0282.0006: input,hidraw1: USB HID v1.11 Keyboard [Razer Razer Huntsman Mini Analog] on usb-0000:00:14.0-6.1.4/input0
input: Razer Razer Huntsman Mini Analog as /devices/pci0000:00/0000:00:14.0/usb3/3-6/3-6.1/3-6.1.4/3-6.1.4:1.1/0003:1532:0282.0007/input/input37
razerkbd 0003:1532:0282.0007: input,hidraw3: USB HID v1.11 Keyboard [Razer Razer Huntsman Mini Analog] on usb-0000:00:14.0-6.1.4/input1
input: Razer Razer Huntsman Mini Analog as /devices/pci0000:00/0000:00:14.0/usb3/3-6/3-6.1/3-6.1.4/3-6.1.4:1.2/0003:1532:0282.0008/input/input38
razerkbd 0003:1532:0282.0008: input,hidraw4: USB HID v1.11 Mouse [Razer Razer Huntsman Mini Analog] on usb-0000:00:14.0-6.1.4/input2
razerkbd 0003:1532:0282.0009: No inputs registered, leaving
razerkbd 0003:1532:0282.0009: hidraw5: USB HID v1.11 Device [Razer Razer Huntsman Mini Analog] on usb-0000:00:14.0-6.1.4/input3
input: Razer Razer Huntsman Mini Analog as /devices/pci0000:00/0000:00:14.0/usb3/3-6/3-6.1/3-6.1.4/3-6.1.4:1.4/0003:1532:0282.000A/input/input39
razerkbd 0003:1532:0282.000A: input,hiddev97,hidraw6: USB HID v1.11 Device [Razer Razer Huntsman Mini Analog] on usb-0000:00:14.0-6.1.4/input4
```

And started the openrazer-daemon:
```
2022-07-23 20:59:06 | razer                          | INFO     | Initialising Daemon (v3.3.0). Pid: 18719
2022-07-23 20:59:06 | razer.screensaver              | INFO     | Initialising DBus Screensaver Monitor
2022-07-23 20:59:06 | razer                          | INFO     | Found device.0: 0003:1532:0282.0008
2022-07-23 20:59:06 | razer.device0                  | INFO     | Initialising device.0 RazerHuntsmanMiniAnalog
2022-07-23 20:59:06 | razer                          | INFO     | Serving DBus
```

And I'm able to run the examples from the examples directory (all seem to work except for the advanced_effect.py)

Do you want me to test anything else before this could get merged? Please let me know if I missed something.

So this was the easy part; now to fix analog input :partying_face:

### CODE REVIEW

**Comment by z3ntu** [comment]
Please also check out a UI like polychromatic or RazerGenie and go through the options shown there and check if they work properly.

Diff looks sane.

**Comment by meeuw** [comment]
:+1: I installed polychromatic and the effects seem to be working (except for the "Inspect Matrix" feature). Not sure if this is expected.

File: driver/razerkbd_driver.c [hunk 1]
```diff
@@ -2983,6 +2997,7 @@ static int razer_kbd_input_mapping(struct hid_device *hdev, struct hid_input *hi
     case USB_DEVICE_ID_RAZER_ORNATA_V2:
     case USB_DEVICE_ID_RAZER_HUNTSMAN_V2:
     case USB_DEVICE_ID_RAZER_HUNTSMAN_V2_ANALOG:
+    case USB_DEVICE_ID_RAZER_HUNTSMAN_MINI_ANALOG:
```
Comment (meeuw) [inline driver/razerkbd_driver.c]:
```suggestion
```

This keyboard doesn't have a volume control knob.

**Comment by meeuw** [comment]
Do you want me to create a new PR for keyboard support in driver mode or should I include it in this PR?

It's my preference to keep PRs small...

**Comment by z3ntu** [comment]
> except for the "Inspect Matrix" feature

How does it not work?

> Do you want me to create a new PR for keyboard support in driver mode or should I include it in this PR?

Definitely a separate one, but I'm also not sure if it's a good idea to include this analog key support here, or if it's a better idea to include this in Linux mainline itself. It goes too much out of scope of openrazer honestly..

**Comment by meeuw** [comment]
> How does it not work?

It might be a bug in polychromatic or I don't undestand how it should work.

I can drag and select "squares" of keys and I'd expect those keys to highlight on my keyboard (it doesn't, keyboard stays dark).

All the effects seem to work, even the "Reactive" one, but I can't select a "Ripple" effect (not available in Polychromatic).

> too much out of scope of openrazer

The required changes for analog support are limited, do you have a clear definition of the scope of openrazer?

**Comment by z3ntu** [comment]
> I can drag and select "squares" of keys and I'd expect those keys to highlight on my keyboard (it doesn't, keyboard stays dark).

Yes, that's how it works with my keyboard. Maybe double check the driver implementation of matrix_custom_frame and matrix_effect_custom to see if there's anything missing, e.g. transaction id or whatnot.

> The required changes for analog support are limited, do you have a clear definition of the scope of openrazer?

Not really but my long term goal is getting rid of the driver completely, so I really don't want to add significant new features to it that are better somewhere else.

**Comment by meeuw** [comment]
Would be awesome if we could upstream the driver although in the past they weren't very enthusiastic about it:

https://www.spinics.net/lists/linux-input/msg58949.html

I wouldn't know an answer to:

> there is quite a struggle to get things to work on all devices correctly

Maybe the drivers can be split into different generations to limit all of the conditions?

I haven't bought this keyboard for the LEDs so fixing matrix_custom_frame isn't much of a priority to me, is this a hard requirement for merging this PR? Maybe I do it later, for now I'd like to focus on analog support.

**Comment by z3ntu** [comment]
The comments upstream were mostly about putting things that you can do in user space (most LED stuff) into the mainline kernel, which I'm also not a fan of because the existing interfaces in the kernel are quite restrictive, compared to our (weird) sysfs file approach in openrazer.

> I haven't bought this keyboard for the LEDs so fixing matrix_custom_frame isn't much of a priority to me, is this a hard requirement for merging this PR? Maybe I do it later, for now I'd like to focus on analog support.

Please spend maybe half an hour on it trying to get it to work. Possibly also MATRIX_DIMS are just wrong, but I don't think I have pcaps for this device to verify what Synapse is doing.

If this doesn't work then drop matrix feature from driver and daemon for this device so we don't show options that don't work.

**Comment by meeuw** [comment]
Right, now I understand that MATRIX_DIMS = matrix dimensions (I was thinking about dimming LEDs :thinking: )

Inspect Matrix is now working on my system!

As for the the "getting rid of the driver completely" I'd love to help with this, do you already have an issue for this?

**Comment by z3ntu** [comment]
> Inspect Matrix is now working on my system!

Perfect, thanks!

> As for the the "getting rid of the driver completely" I'd love to help with this, do you already have an issue for this?

Some work was started a long time ago in #382 , and also by now years ago I started a complete rewrite with https://github.com/z3ntu/razer_test but it has stalled due to lack of time really.

But generally for doing this (and rolling it out) I require some sort of integration test where all of the devices are kind of virtualized to see if each device is still getting the same packets as before, to minimize the regression potential. I really don't want to have to fix up a big number of our 150+ supported devices after rolling out such a change..

**Review by z3ntu — APPROVED** [review]
LGTM, thanks for the PR!

**Pull request merged**